### PR TITLE
[JAX] Activation/Normalization to output amax for later quantization in CurrentScaling

### PR DESCRIPTION
**Pull request by phu0ngng**
# Description

Activation/Normalization to output amax for later quantization in CurrentScaling

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: transformer_engine/jax/cpp_extensions/activation.py [hunk 1]
```diff
@@ -148,36 +148,43 @@ class ActLuPrimitive(BasePrimitive):
...
+        amax_scope,
+        transpose_batch_sequence,
+        output_amax_when_no_scaling,
+        is_outer,
     ):
         """
         te_act_lu_p abstract
         """
-        del act_enum, act_params
+        del act_enum, act_params, amax_scope, transpose_batch_sequence, output_amax_when_no_scaling
```
Comment (jberchtold-nvidia) [inline transformer_engine/jax/cpp_extensions/activation.py]: Can we add an assert here that we are using `ScalingMode.NO_SCALING` if `output_amax_when_no_scaling` is true?

File: transformer_engine/jax/cpp_extensions/activation.py [hunk 2]
```diff
@@ -452,26 +495,40 @@ def partition(
...
+                is_outer=True,
             )
 
             if scaling_mode == ScalingMode.DELAYED_TENSOR_SCALING.value:
-                global_updated_amax = all_reduce_max_along_all_axes_except_PP(local_amax, mesh)
+                global_updated_amax = all_reduce_max_along_all_axes_except_PP(
+                    local_updated_amax, mesh
+                )
+            elif scaling_mode == ScalingMode.NO_SCALING.value and output_amax_when_no_scaling:
+                global_updated_amax = amax_scope.all_reduce_amax_along_TPSP_and_FSDP(
```
Comment (jberchtold-nvidia) [inline transformer_engine/jax/cpp_extensions/activation.py:531]: Thanks for refactoring this so the logic for amax reduction isn't duplicated between primitives!

File: transformer_engine/jax/cpp_extensions/activation.py [hunk 3]
```diff
@@ -1232,19 +1400,22 @@ def quantize_dact_dbias(
         f" {x.shape} and act_len {act_len}"
     )
 
-    scale = jnp.empty((), jnp.float32)
+    scale = jnp.empty((1,), jnp.float32)
+    amax = jnp.zeros((1,), jnp.float32)  # need to init with zero and shape=(1,)
```
Comment (jberchtold-nvidia) [inline transformer_engine/jax/cpp_extensions/activation.py:1411]: Why does this amax need to be an input to the primitive? We only every output an amax from fused act, right?  Is this just to allow JAX to zero out the buffer instead of nvte_memset?

File: transformer_engine/jax/cpp_extensions/normalization.py [hunk 4]
```diff
@@ -110,15 +111,20 @@ def abstract(
...
         scale_dtype,
+        amax_scope,
+        transpose_batch_sequence,
+        output_amax_when_no_scaling,
         is_outer,
     ):
         """
         LayerNorm fwd inner primitive abstract
         """
+        del amax_scope, transpose_batch_sequence, output_amax_when_no_scaling
```
Comment (jberchtold-nvidia) [inline transformer_engine/jax/cpp_extensions/normalization.py]: Can we assert here in the abstract that `output_amax_when_no_scaling` is only allowed to be True when the scaling mode is NO_SCALING and we are not using cuDNN norm? Same for norm bwd abstract

File: transformer_engine/jax/csrc/extensions/normalization.cpp [hunk 5]
```diff
@@ -123,8 +132,6 @@ Error_Type NormForwardFFI(cudaStream_t stream, Buffer_Type x_buf, Buffer_Type sc
 
   if (scaling_mode == JAXX_Scaling_Mode::DELAYED_TENSOR_SCALING && is_fp8_dtype(out_dtype)) {
     output_tensor.set_scale(scale, DType::kFloat32, std::vector<size_t>{1});
-    nvte_memset(amax, 0, sizeof(float), stream);
```
Comment (jberchtold-nvidia) [inline transformer_engine/jax/csrc/extensions/normalization.cpp:126]: Do we know if `jnp.zeros` is faster or is this just a simplification for refactoring purposes? Previously, we saw a cudaMemset being slower than a fill kernel, but I am unsure which `jnp.zeros` is doing or if JAX/XLA can combine multiple zeros into a single op.

File: transformer_engine/jax/flax/module.py
Comment (jberchtold-nvidia) [inline transformer_engine/jax/flax/module.py:1081]: Can we add a TODO to utilize `output_amax_if_no_scaling` here for the unfused norm codepath in the modules? I agree, that it is not as high priority as our fused VJP path as that is the most optimal path we want to optimize

Comment (jberchtold-nvidia) [reply transformer_engine/jax/flax/module.py:1081]: Update: I think `output_amax_if_no_scaling` should not be in the public API and should instead be internal to `cpp_extensions`. In the future, when we want to support this unfused cases like here, we can add it. But for now we don't, so we don't need to do anything here

File: transformer_engine/jax/cpp_extensions/normalization.py [hunk 6]
```diff
@@ -1091,7 +1178,9 @@ def rmsnorm_fwd(
     zero_centered_gamma: bool,
     epsilon: float,
     quantizer: Optional[Quantizer],
-    amax_scope: AmaxScope = AmaxScope.LOCAL,
+    amax_scope: AmaxScope = AmaxScope.TPSP,
+    transpose_batch_sequence: bool = False,
+    output_amax_when_no_scaling: bool = False,
```
Comment (jberchtold-nvidia) [inline transformer_engine/jax/cpp_extensions/normalization.py:1190]: I don't think `output_amax_when_no_scaling` should be exposed publicly outside of `cpp_extensions` for now as afaict we don't use it anywhere in this PR. The logic to use this can be determined internally in `cpp_extensions` as is already done for the current scaling case. If I'm missing something, lmk, thanks!

File: transformer_engine/jax/cpp_extensions/activation.py [hunk 7]
```diff
@@ -1209,6 +1374,9 @@ def quantize_dact_dbias(
     is_dbias: bool = True,
     quantizer: Optional[Quantizer] = None,
     act_params: Optional[ActivationParams] = None,
+    amax_scope: AmaxScope = AmaxScope.LOCAL,
+    transpose_batch_sequence: bool = False,
+    output_amax_when_no_scaling: bool = False,
```
Comment (jberchtold-nvidia) [inline transformer_engine/jax/cpp_extensions/activation.py:1386]: I don't think `output_amax_when_no_scaling` should be exposed publicly outside of `cpp_extensions` for now as afaict we don't use it anywhere in this PR. The logic to use this can be determined internally in `cpp_extensions` as is already done for the current scaling case. If I'm missing something, lmk, thanks!

File: transformer_engine/jax/cpp_extensions/normalization.py [hunk 8]
```diff
@@ -882,6 +937,8 @@ def layernorm_fwd(
     epsilon: float,
     quantizer: Optional[Quantizer],
     amax_scope: AmaxScope = AmaxScope.LOCAL,
+    transpose_batch_sequence: bool = False,
+    output_amax_when_no_scaling: bool = False,
```
Comment (jberchtold-nvidia) [inline transformer_engine/jax/cpp_extensions/normalization.py:948]: I don't think `output_amax_when_no_scaling` should be exposed publicly outside of `cpp_extensions` for now as afaict we don't use it anywhere in this PR. The logic to use this can be determined internally in `cpp_extensions` as is already done for the current scaling case. If I'm missing something, lmk, thanks!

File: transformer_engine/jax/cpp_extensions/activation.py (hunk 3 above)
Comment (phu0ngng) [reply transformer_engine/jax/cpp_extensions/activation.py:1411]: Yes, I added amax to activation and normalization primitives for two reasons:
1. Eliminate the nvte_memset
2. Stay consistent with quantization primitives.

File: transformer_engine/jax/cpp_extensions/normalization.py (hunk 4 above)
Comment (phu0ngng) [reply transformer_engine/jax/cpp_extensions/normalization.py]: Done for NormFWD. We don't quantize the NormBWD output thus this change is irrelevant.

File: transformer_engine/jax/csrc/extensions/normalization.cpp (hunk 5 above)
Comment (phu0ngng) [reply transformer_engine/jax/csrc/extensions/normalization.cpp:126]: XLA should be able to combine zero initialization.

File: transformer_engine/jax/flax/module.py
Comment (phu0ngng) [reply transformer_engine/jax/flax/module.py:1081]: It's probably tricky if we want to support this unfuse case even in the future because if you output the amax in the fwd, you will need to deal with its gradient in the bwd.
In fact, we have this unfuse code path here because the VJPs did not support BF16 in the past. Now it is, so we should be able to refactor the code to use VJP.

File: transformer_engine/jax/cpp_extensions/normalization.py (hunk 8 above)
Comment (phu0ngng) [reply transformer_engine/jax/cpp_extensions/normalization.py:948]: I expose `output_amax_when_no_scaling` in `normalization.py` and `activation.py` because we recursively call the function inside the function itself, for example, here https://github.com/phu0ngng/TransformerEngine/blob/main/transformer_engine/jax/cpp_extensions/normalization.py#L950-L957. 

In general, I don't anticipate users using cpp_extensions functions as they are not differential. I would be more cautious if we need to change VJP APIs.

Comment (phu0ngng) [reply transformer_engine/jax/cpp_extensions/normalization.py:948]: Another possible way is to split the implementation into an internal `_impl` function, as we did in the `_quantize_dbias_impl`. Let me add a todo for that.

Comment (jberchtold-nvidia) [reply transformer_engine/jax/cpp_extensions/normalization.py:948]: I see, yes I think an internal `_impl` method is a good idea for this but just adding a TODO in this PR sounds good to me.

**Comment by phu0ngng** [comment]
Pipeline 36141587 passed.
